packages/persistent-merkle-tree/src/hashComputation.ts

```diff
@@ -0,0 +1,216 @@
+import type {Node} from "./node";
+
+/**
+ * HashComputation to be later used to compute hash of nodes from bottom up.
+ * This is also an item of a linked list.
+ *     ╔═════════════════════╗             ╔══════════════════════╗
+ *     ║       dest          ║             ║      next_dest       ║
+ *     ║     /      \        ║  ========>  ║     /       \        ║
+ *     ║  src0      src1     ║             ║  next_src0  next_src1║
+ *     ╚═════════════════════╝             ╚══════════════════════╝
+ */
+export type HashComputation = {
+  src0: Node;
+  src1: Node;
+  dest: Node;
+  next: HashComputation | null;
+};
+
+/**
+ * Model HashComputation[] at the same level that support reusing the same memory.
+ * Before every run, reset() should be called.
+ * After every run, clean() should be called.
+ */
+export class HashComputationLevel {
+  private _length: number;
+  private _totalLength: number;
+  // use LinkedList to avoid memory allocation when the list grows
+  // always have a fixed head although length is 0
+  private head: HashComputation;
+  private tail: HashComputation | null;
+  private pointer: HashComputation | null;
+
+  constructor() {
+    this._length = 0;
+    this._totalLength = 0;
+    this.head = {
+      src0: null as unknown as Node,
+      src1: null as unknown as Node,
+      dest: null as unknown as Node,
+      next: null,
+    };
+    this.tail = null;
+    this.pointer = null;
+  }
+
+  get length(): number {
+    return this._length;
+  }
+
+  get totalLength(): number {
+    return this._totalLength;
+  }
+
+  /**
+   * run before every run
+   */
+  reset(): void {
+    // keep this.head object, only release the data
+    this.head.src0 = null as unknown as Node;
+    this.head.src1 = null as unknown as Node;
+    this.head.dest = null as unknown as Node;
+    this.tail = null;
+    this._length = 0;
+    // totalLength is not reset
+    this.pointer = null;
+  }
+
+  /**
+   * Append a new HashComputation to tail.
+   * This will overwrite the existing HashComputation if it is not null, or grow the list if needed.
+   */
+  push(src0: Node, src1: Node, dest: Node): void {
+    if (this.tail !== null) {
+      let newTail = this.tail.next;
+      if (newTail !== null) {
+        newTail.src0 = src0;
+        newTail.src1 = src1;
+        newTail.dest = dest;
+      } else {
+        // grow the list
+        newTail = {src0, src1, dest, next: null};
```

Comment:
so this part optimizes memory over the array implementation?

Reply:
yes, when we are at length 1M, for example, pushing a new item cause very small allocation for LinkedList while could be a big deal for an array, especially array of objects

that's why we see a lot of 1s spikes using Array

Reply:
As far as I know, v8 `Array` is like an object, which shouldn't incur a large allocation when new properties are added.

So I'm not really sure why this works from a theory perspective, but the data looks promising, and can't argue with that :shrug:

Reply:
yes I tried to answer the same question, looks like when adding a new object to Array, if it reaches its internal memory it needs to allocate and copy all items to a new space. This response from ChatGPT looks reasonable to me:
```
JavaScript arrays are dynamically sized, but internally, they need to manage memory blocks to store the array elements:

Allocation on Growth: If the existing capacity of the array is insufficient to accommodate a new element, the JavaScript engine must allocate a new, larger block of memory to hold the array's contents and copy existing elements to this new space. This operation increases temporary memory usage and may trigger GC if the memory is tight.
Fragmentation: Repeatedly resizing arrays can lead to memory fragmentation, which might prompt the garbage collector to run more frequently to reclaim unused memory segments.
```